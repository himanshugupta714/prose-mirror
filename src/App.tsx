import useProseMirror from "./hooks/useProse"
import Editor, { Handle } from "./ProseMirror"

import { schema } from "prosemirror-schema-basic"
import { TextSelection } from "prosemirror-state"
import { useCallback, useMemo, useRef } from "react"

function App() {
  const [state, setState] = useProseMirror({
    schema,
    doc: schema.node(
      "doc",
      {},
      schema.node("paragraph", {}, schema.text("12 345 678 910"))
    ),
  })

  const ref = useRef<Handle>(null)

  const handleSelection = useCallback(() => {
    const { from, to } = state.selection

    let startPos = from // The start position of the selection
    let endPos = to - 1 // The end position of the selection

    if (endPos - startPos > 1) {
      const nodeBeforeText = state.doc.resolve(startPos).nodeBefore
      const nodeAfterText = state.doc.resolve(endPos).nodeAfter

      const nodeBeforeWordsArray = nodeBeforeText?.text?.split(" ")
      const nodeAfterWordsArray = nodeAfterText?.text?.split(" ")

      const lastWordBefore = nodeBeforeWordsArray?.pop()
      const lastWordAfter = nodeAfterWordsArray.shift()

      const updatedStart = startPos - lastWordBefore?.length
      const updatedEnd = endPos + lastWordAfter?.length

      if (ref.current?.view) {
        ref.current.view.dispatch(
          state.tr.setSelection(
            new TextSelection(
              state.doc.resolve(updatedStart),
              state.doc.resolve(updatedEnd)
            )
          )
        )
      }
    }
  }, [state.selection, state.doc, state.tr])

  const handleDOMEvents = useMemo(
    () => ({
      mouseup: handleSelection,
    }),
    [handleSelection]
  )

  return (
    <Editor
      ref={ref}
      handleDOMEvents={handleDOMEvents}
      state={state}
      onChange={setState}
    />
  )
}

export default App
