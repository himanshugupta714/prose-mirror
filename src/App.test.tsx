import { render } from "@testing-library/react"
import App from "./App"

describe("App Container", () => {
  test("App", () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })
})
