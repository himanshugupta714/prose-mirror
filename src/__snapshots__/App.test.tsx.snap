// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Container App 1`] = `
<div>
  <div>
    <div
      class="ProseMirror"
      contenteditable="true"
      translate="no"
    >
      <p>
        12 345 678 910
      </p>
    </div>
  </div>
</div>
`;
